twitter_scrooge/twitter_scrooge.bzl

```diff
@@ -0,0 +1,274 @@
+    fail('scrooge_srcjar target must depend on scrooge_srcjar targets sufficient to ' +
+         'cover the transitive graph of thrift files. Uncovered sources: ' + missing)
+
+def _gen_scrooge_srcjar_impl(ctx):
+  # These are the thrift sources whose generated code we will "own" as a target
+  immediate_thrift_srcs = _collect_immediate_srcs(ctx.attr.deps)
+
+  # These are the thrift sources in the dependency graph. They are necessary
+  # to generate the code, but are not "owned" by this target and will not
+  # be in the resultant source jar
+
+  transitive_thrift_srcs = _collect_transitive_srcs(ctx.attr.deps)
+
+  only_transitive_thrift_srcs = set()
+  for src in transitive_thrift_srcs:
+    if src not in immediate_thrift_srcs:
+      only_transitive_thrift_srcs += [src]
+
+  transitive_owned_srcs = _collect_owned_srcs(ctx.attr.deps)
+
+  _assert_set_is_subset(only_transitive_thrift_srcs, transitive_owned_srcs)
+
+  #TODO would there be a case in which we want/need to allow more jars here?
+  transitive_cjars = [ctx.files._scrooge_core, ctx.files._libthrift, ctx.files._scalalib]
+  transitive_cjars = [file for files in transitive_cjars for file in files]
+
+  cjars = ctx.files._scrooge_generator + \
+          ctx.files._scopt + \
+          ctx.files._util_core + \
+          ctx.files._util_logging + \
+          ctx.files._scala_parser_combinators + \
+          transitive_cjars
+  write_manifest(ctx) #TODO do we need something like this?
+
+  cmd = """
+rm -rf {out}_tmp
+set -e
+{java} -classpath "{jars}" com.twitter.scrooge.Main -d {out}_tmp --gen-file-map {gen_file_map} $@
+find {out}_tmp -exec touch -t 198001010000 {{}} \;
+touch -t 198001010000 {manifest}
+{jar} cmf {manifest} {out} -C {out}_tmp .
+rm -rf {out}_tmp
+""".format(
+    java=ctx.file._java.path,
+    out=ctx.outputs.srcjar_polluted.path,
+    manifest=ctx.outputs.manifest.path,
+    jar=ctx.file._jar.path,
+    jars=":".join([f.path for f in cjars]),
+    gen_file_map=ctx.outputs.gen_file_map.path,
+  )
+
+  ctx.action(
+    inputs=list(transitive_thrift_srcs) +
+        list(transitive_owned_srcs) +
+        cjars +
+        ctx.files._jdk +
+        ctx.files._scalasdk +
+        [ctx.outputs.manifest, ctx.file._jar],
+    outputs=[ctx.outputs.srcjar_polluted, ctx.outputs.gen_file_map],
+    command=cmd,
+    progress_message="scrooge generation %s" % ctx.label,
+    # Since we have access to the graph, we don't have to muck with the
+    # thrift_library tar, we can just use the sources directly
+    arguments=[f.path for f in transitive_thrift_srcs],
+  )
+
+  pluck=ctx.executable._pluck_scrooge.path
+
+  pluck_cmd="""
+rm -rf {out}_tmp_polluted
+rm -rf {out}_tmp
+set -e
+mkdir {out}_tmp_polluted
+mkdir {out}_tmp
+unzip -o {polluted} -d {out}_tmp_polluted 2>/dev/null
+touch -t 198001010000 {manifest}
+touch {out}_tmp_polluted/gen_pluck_cmd
+{pluck} {jar} {manifest} {out} {out}_tmp_polluted/gen_pluck_cmd {out}_tmp_polluted {genmap} {include_base} {pluck_base} $@
```

Comment:
I think it is fine to merge using python, but actually due to bazelbuild/bazel#890, we are hoping that there is a system python and it is the right version.

This is fine for now, but rewriting that code in scala should not be hard (you can even make the jar without shelling out then: https://docs.oracle.com/javase/7/docs/api/java/util/jar/package-summary.html